### ISSUE TITLE
Hooks

### ISSUE THREAD

**Issue by peXed** [gridsome/gridsome#974]
## Summary

Hooks like onCreatePage, onCreateNode, etc.

### Basic example

Gridsome is missing a hook system like in Gatsby. It can be used in many different ways.

### Motivation

A big use case would be l18n. You can very easly create a system, that is "duplicating" a page on creation with a swapped environment variable and path, to create different translations and a very good translation hierachy.

**Comment by thetre97** [comment]
@peXed Gridsome does indeed have these API methods? Check out https://gridsome.org/docs/server-api/, https://gridsome.org/docs/server-api/#apioncreatenodefn for more information...

**Comment by peXed** [comment]
@thetre97  thanks for the fast reply!

You are right, onCreateNode is already implemented. But unfortunatly, the onCreatePage isn't. Or is it not documented?

**Comment by noxify** [comment]
@peXed it seems only `onCreateNode` is implemented currently.

Based on the code, the current hooks are currently implemented:
https://github.com/gridsome/gridsome/blob/9ed78c98958b6108cff95afcd97b6e87d8fc8184/gridsome/lib/app/PluginAPI.js#L83

**Comment by thetre97** [comment]
@peXed Just for my own interest, what sort of things would you use `onCreatePage` for?

**Comment by peXed** [comment]
@thetre97 I would copy each page, for every language I need. Change the path (add the language at the start of the path) and inject a environment variable, or if possible metavalue.
With that variable, the content gets translated.

At the moment, I solve that issue with a gulp build process and generate the website multiple times (with a changed environment variable and basePath) and copy them into each other. But with that approach, I have a lot of duplicated assets and g-link for switching the language won't work.

**Comment by courcelan** [comment]
@peXed I believe you can do this in `gridsome.server.js` as is.

```javascript
api.createPages(
    async ({ createPage }) => {
        // ... load initial page content/context ...

        // ... create page 1 ...
        createPage(
            path,
            component,
            context,
            queryVariables
        )       

        // ... create page alternate (say for spanish)...
        createPage(
            path: `es/${ path }`,
            component,
            context,
            queryVariables
        )       

        // ... etc ...
    }
)
```

https://gridsome.org/docs/pages-api/#create-a-page

maybe I'm reading your request wrong, but you can do a lot in `api.createPages` and using the context prop to `createPage` you can really load up the component as needed.

**Comment by peXed** [comment]
@courcelan Yes, you can create the pages. But at the moment, there is no way to get all pages created by Gridsome (internally).

The hook would basically getting called whenever a page gets created (by Gridsome or a Plugin or custom code, that is using the createPages API). This hook could perfectly be used to "catch" all pages and create the duplicates for the different languages.

**Comment by courcelan** [comment]
ah gotcha

**Comment by hjvedvik** [comment]
I'm working on a `onCreatePage` hook now, and it would be nice with some feedback on the implementation. I'm thinking about making it very similar to the `onCreateNode` hook. Where you can return `null` to exclude the node or return a new one to modify it. But the `onCreatePage` will also allow returning an array of new page options to convert a single page into several new ones.

```js
api.onCreatePage(options => {
  // removes the page
  return null

  // modify the page
  return { ...options, path: '/another-path' }
  // or by mutating the options
  options.path = '/another-path'

  // convert the page into multiple pages
  return [
    { ...options, path: '/page-1' },
    { ...options, path: '/page-2' }
  ]

  // return nothing or `undefined` to skip
})
```

Another alternative is to make it more like how it works in Gatsby.

```js
api.onCreatePage((options, { removePage, createPage }) => {
  // removes the page
  removePage(options)

  // modify the page by removing it and creating a new one
  removePage(options)
  createPage({ ...options, path: '/another-path' })
  // or by mutating the options
  options.path = '/another-path'

  // convert the page into multiple pages
  removePage(options)
  createPage({ ...options, path: '/page-1' })
  createPage({ ...options, path: '/page-2' })
})
```

The first alternative might be easier to use but is limited to the current options in the hook. While the second one will let you remove other pages as well (if there are any use cases for doing so?). And it looks more like how things are done in the `api.createPages()` hook. What do you think? @peXed @courcelan @thetre97

**Comment by courcelan** [comment]
@hjvedvik Definitely excited for this. Thanks for the inclusion!

I like the explicitness of `removePage` / `createPage` and I do like the lean power of the `onCreateNode` style. Ultimately, i think going with the Gatsby style is more straightforward and should be leaned towards. I'm thinking primarily about: 

1) The confusion in version 1 around the many ways to return a page value or not. By your example, there are at least four results/paths that I can take. version 2 doesn't return anything but nicely acts as a gateway into the api methods. While I like the ability to mutate options as a sort of "escape hatch", it seems slightly dangerous if a developer does some crazy involved functionality within the hook. I remember having something like this in Vuex-ORM and burning myself a few times with the mutation, or at least (if I'm honest!), not being perceptive enough to understand how to use it correctly!

2) I think `onCreatePage` is going to be used frequently, especially for instances like i18n production. I can easily see the questions that are generated by developers checking it out and I think the version 2 explicitness is more easily maintained in everyone's heads and certainly can be defined through clear, approachable documentation.

I feel like the use case for removing other pages not associated with the current page is at the edge of my brain, I can't really give an example at present but I'm sure its there! `removePage`/`createPage` seems certainly more open-ended for extension with its use, while the returning of an array (version 1) at best, leaves a developer to think only associated pages should be created (nor could you remove "random" pages, and at worst, remains pretty opaque.

Cheers

**Comment by peXed** [comment]
@hjvedvik Thank you, for you great work!

I'm on @courcelan side. I think the second approach is cleaner and more API style. It would be awesome if this API gets exposed to other places as well (e.g. other hooks).

**Comment by hjvedvik** [comment]
Yes, I agree  :) I have the second approach working now. But there are a few things I need to sort out before it's ready. And I might do the same with `onCreateNode` and deprecate returning anything from it.

@peXed Regarding other hooks, while having some "main" hooks like `onBootstrap`, `onCreateNode`, `onCreatePage` etc, the current plan is to use the [`tapable`](https://github.com/webpack/tapable) package more internally to let plugins hook into other parts in core. Gridsome has a few experimental hooks atm but we'll probably add many more once the internal APIs are more stable. But I'm open to other alternatives.

**Comment by thetre97** [comment]
@hjvedvik Incidentally, talking about the `onBootstrap` hook, I could only get this to work if I changed it to `tapPromise`, as opposed to `tapAsync`? If using `tapAsync`, the execution just freezes. Looking at the tapable documentation, it seems that with Async, you should pass a callback function, but I couldn't trace the gridsome hooks back to see if this was being passed or not - it seems not as the parameters from that hook are undefined..

Perhaps I should open a separate issue/PR 😏
